streaming/src/test/scala/org/apache/spark/streaming/scheduler/RateControllerSuite.scala

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.streaming.scheduler
+
+import scala.collection.mutable
+import scala.reflect.ClassTag
+import scala.util.control.NonFatal
+
+import org.scalatest.Matchers._
+import org.scalatest.concurrent.Eventually._
+import org.scalatest.time.SpanSugar._
+
+import org.apache.spark.streaming._
+import org.apache.spark.streaming.scheduler.rate.RateEstimator
+
+class RateControllerSuite extends TestSuiteBase {
+
+  override def actuallyWait: Boolean = true
+
+  test("rate controller publishes updates") {
+    val ssc = new StreamingContext(conf, batchDuration)
+    val dstream = new MockRateLimitDStream(ssc, Seq(Seq(1)), 1)
+    val output = new TestOutputStreamWithPartitions(dstream)
```

Comment:
Good point about leaking contexts, I'll see what I can do. Since `runStreams` stops the context, we'll stop it twice (I know, it only logs a warning, but still it seems suboptimal).

I'd like to keep using `runStreams`, otherwise I'll have to duplicate some of its implementation. The output testing is collateral, but not bad in itself, at least not worse than duplicating the logic around manual clock advancing. I *do* need to simulate running for a few batch intervals to see how the rate gets updated. 

Lastly, this *particular* test doesn't need it, but I updated to be similar to the others here, and I think it "ties the suit together" :). If you insist I can revert it, but the other two tests, that actually run the whole pipeline to see rate updates in action, still need `runStreams`